### ISSUE TITLE
Resource management

### ISSUE THREAD

**Issue by CoolRustacean** [ziglang/zig#7992]
Currently, the Zig convention for resource management is a form of RAII. However, one of tenets of RAII is that resources should only be held while the object is alive. Sometimes thats's easy. For example, in a linked list, each node owns the next node. Therefore, each node should deinitialize the next one.
However, deallocating an object of the type `AutoHashMap(u32, ArrayList(u8))`, for example, requires iterating through the entire `HashMap` and calling `deinit()` on each `ArrayList`.

This breaks RAII.
A function that returns an `AutoHashMap(u32, ArrayList(u8))` requires the caller to deinitialize every `ArrayList(u8)` before deinitializing the whole hashmap.

Maybe we could introduce a `Deinit` interface, kind of like the `Allocator` interface. Thus, if type `T` implements `Deinit`, calling `deinit()` on an `ArrayList(T)` will call `deinit()` on every item in the `ArrayList`.

**Comment by Tetralux** [comment]
If you have an `ArrayList(*MyType)`, then the list doesn't own those `MyType`s.

Zig uses a pretty _manual_ form of RAII, to the extent that you could call it that. No auto-destruction, for instance.

Zig is designed to help you maintain awareness about what precisely you're asking the computer to do, and do it efficiently.
Deinitting every element in an arraylist is generally not an efficient operation, but also isn't always what you actually want; so if you really want that, you have to explicitly say so.

"What if I want to destroy the list, and everything in it?"
"If you _really_ want to: just `for (list.items) |*e| e.deinit();`."


This also gives you a prod into thinking that you maybe should be using a different allocation strategy; maybe even, not bothering to free them in the first place, for example.

"Can I use a arena for this list so that I can free everything without having to iterate?"
"Is the program about to exit, anyway?"

**Comment by SpexGuy** [comment]
I think much of the community, myself included, would dispute this premise.

> Currently, the Zig convention for resource management is a form of RAII

RAII is something of an ambiguous term, but in C++ it means that resource acquisition is tied to object lifetimes.  Zig doesn't have a concept of object lifetimes, so it can't have RAII in that sense.  But also, there's nothing preventing you from reacquiring resources after they have been released for a single object in Zig.

Another important break is that Zig allows you to copy management structures by value.  You can return a std.ArrayList by value, for example.  We can do this without adding all of the complexity of rvalue references to the language specifically because we don't have RAII in the language.

> For example, in a linked list, each node owns the next node. Therefore, each node should deinitialize the next one.

Most linked lists should not work like this.  It would be disastrous for performance, except in very specialized cases.  One common use case that violates this pattern is a linked list used to track adjacency through a list of edges, for example in a clipping algorithm.  The edges are all stored in a contiguous list, and linked lists are woven through that list for different traversal orders.  A generic linked list class making this choice about deinitialization heavily limits how useful it is.  Zig avoids this limitation by not tying resource cleanup to memory lifetime.  Data structures never attempt to initialize or deinitialize objects.  Thus we have a pattern of "whoever init()s also deinit()s".  This is symmetrical, and makes sense for most use cases.  Transferring ownership is a difficult pattern to understand, it should be avoided if possible.

Similarly, it doesn't make sense to tie deinitialization to a type.  If a type implements deinit, but you want to retain ownership in some cases, then you would be unable to use that type with data structures which call deinit in those cases.

---

As always, you can choose to make and use this convention in your own projects.  But the standard lib should not use this convention.  It reduces reusability of data structures.

**Comment by CoolRustacean** [comment]
> If you have an `ArrayList(*MyType)`, then the list doesn't own those `MyType`s.

Granted that in an `ArrayList(*MyType)`, the `ArrayList` is not the owner. However, in an `ArrayList(MyType)`, the `ArrayList` _is_ the owner. Any attempt to borrow or share a `MyType` in the `ArrayList` would either create a shallow copy or create a dangling pointer.
```zig
// ...

var array = std.ArrayList(MyType).init(allocator);
defer array.deinit();

try array.append(MyType.new());
var ptr = &array.items[0]; // 'ptr' should not live longer than 'array'
```

> Zig uses a pretty manual form of RAII, to the extent that you could call it that. No auto-destruction, for instance.

I know.
> "What if I want to destroy the list, and everything in it?"
> "If you _really_ want to: just `for (list.items) |*e| e.deinit();`."

That's exactly my point. It makes the caller responsible for calling `deinit()` on objects that it doesn't own. What happened to RAII and code reuse?

**Comment by MasterQ32** [comment]
> What happened to RAII [...]?

RAII is not a language feature in Zig, but a user space convention not enforced by the standard library. RAII requires a lot more than just constructors and destructors and makes a lot of stuff way more complex.

> Any attempt to borrow or share a MyType in the ArrayList would either create a shallow copy or create a dangling pointer.

RAII wouldn't solve this problem at all. It's something you have to take care of as a programmer, the language won't help you with that

**Comment by CoolRustacean** [comment]
> I think much of the community, myself included, would dispute this premise.
> 
> > Currently, the Zig convention for resource management is a form of RAII
> 
> RAII is something of an ambiguous term, but in C++ it means that resource acquisition is tied to object lifetimes. Zig doesn't have a concept of object lifetimes, so it can't have RAII in that sense. But also, there's nothing preventing you from reacquiring resources after they have been released for a single object in Zig.
> 
> Another important break is that Zig allows you to copy management structures by value. You can return a std.ArrayList by value, for example. We can do this without adding all of the complexity of rvalue references to the language specifically because we don't have RAII in the language.
> 
> > For example, in a linked list, each node owns the next node. Therefore, each node should deinitialize the next one.
> 
> Most linked lists should not work like this. It would be disastrous for performance, except in very specialized cases. One common use case that violates this pattern is a linked list used to track adjacency through a list of edges, for example in a clipping algorithm. The edges are all stored in a contiguous list, and linked lists are woven through that list for different traversal orders. A generic linked list class making this choice about deinitialization heavily limits how useful it is. Zig avoids this limitation by not tying resource cleanup to memory lifetime. Data structures never attempt to initialize or deinitialize objects. Thus we have a pattern of "whoever init()s also deinit()s". This is symmetrical, and makes sense for most use cases. Transferring ownership is a difficult pattern to understand, it should be avoided if possible.
> 
> Similarly, it doesn't make sense to tie deinitialization to a type. If a type implements deinit, but you want to retain ownership in some cases, then you would be unable to use that type with data structures which call deinit in those cases.
> 
> As always, you can choose to make and use this convention in your own projects. But the standard lib should not use this convention. It reduces reusability of data structures.


RAII and performance aside, wouldn't you agree that a resource bound to an object should not outlive the object itself? There is no well-formed use of deinitializing an object and then using the resource.

**Comment by CoolRustacean** [comment]
> RAII requires a lot more than just constructors and destructors and makes a lot of stuff way more complex.

How so?
> RAII wouldn't solve this problem at all. It's something you have to take care of as a programmer, the language won't help you with that

No. If the standard library were RAII-compliant, calling `deinit()` on an object should deinitialize all resources that the object contains. Calling `deinit()` on a binary tree should deinitialize the entire tree, not just the first node.

**Comment by CoolRustacean** [comment]
> Another important break is that Zig allows you to copy management structures by value. You can return a std.ArrayList by value, for example. 

Why would you ever want to do _that_? Use `*std.ArrayList` for granting mutable access, and `*const std.ArrayList` for immutable access. Shallow copying an object containing a resource is a bad idea.

**Comment by MasterQ32** [comment]
> Why would you ever want to do _that_? Use `*std.ArrayList` for granting mutable access, and `*const std.ArrayList` for immutable access. Shallow copying an object containing a resource is a bad idea.

Simple: It would require you to allocate that ArrayList to return a pointer. Otherwise you return a pointer to stack memory which is illegal ;)

Also nothing in the ArrayList points to itself, so it's not a problem copying it around like any other block of memory, it doesn't need to have a fixed address

**Comment by zzyxyzz** [comment]
@CoolRustacean,

I share your annoyance at having to iterate over and manually deallocate elements in a container, but this is not something that is easily fixed without introducing other problems. There are two solutions one could try: 1) full-scale lifetime-based resource management and 2) changing the convention of the standard library from shallow to deep `deinit()`. Here is why both wouldn't work in Zig, even though they do in other languages:

 1. Lifetime-based resource management.

If you want to follow in the footsteps of C++ and  automatically insert destructor calls when an object goes out of scope, you will eventually have to introduce the whole shebang of constructors, destructors, copy constructors, move constructors, r-value references and the associated syntax. All of these things exist for a reason in C++ and it is hard to see how Zig could avoid them. Getting away from all that stuff was probably was one of the main motivations for Zig in the first place, so I have a strong hunch that full-scale RAII is completely off the table.

 2. Recursive deinit convention.

To be clear, this is supposed to mean that `deinit()` is never invoked automatically by the compiler, but the `deinit()` methods of `ArrayList` etc. are written to call `deinit()` on their children before freeing the underlying memory.

The first problem was already pointed out by others, namely that you lose control and flexibility. Even if you don't care about that, there are many subtle problems with recursive `deinit()`. First off, `ArrayList(T)` is a generic type, so to perform a deep deinit, it needs to be able to call `.deinit()` on T regardless of what T is, including builtin types, pointers, etc. This circumstance pretty much requires that all types -- builtin or custom -- be given a default implementation of `deinit()` by the compiler. For elementary types and pointers it would obviously be a no-op, but how it should work with custom types is less clear. One could be tempted to say that the default destructor should call deinit on all struct fields, but then we are faced with the problem of what to do when the destructor is overridden:

```zig
const MyType1 = struct {
    buffer: ArrayList(u32); // default destructor deinits "owned" objects
    resource: *Foo;         // ... but doesn't touch pointers
};

const MyType2 = struct {
    buffer: ArrayList(u32);
    resource: *Foo;

    fn deinit(self: *MyType2) {
        resource.deinit();  // explicit destructor deinits pointed-to resource
        // ... but what happens to buffer?
    }
};
```

If we say that an explicit `deinit` method completely overrides the default one, that looks like it will be a major source of errors, if people add an explicit constructor to deal with pointers and other stuff, but forget that they now need to deal with deiniting values too. If, on the other hand, we say that the explicit destructor is "on top" of the default one, that's even worse, because it may cause the deinit method of a member to be called twice, and in addition locks *all* code into using the recursive deinit semantics, with no reasonable escape hatch.

Both options are bad, so the only remaining thing is to say that all default destructors do absolutely nothing (even on structs). This is still a workable solution for the original problem, except now we may forget to deallocate "owned" resources. If we forget to implement a `deinit()` method, then the compiler will not warn us, but silently generate an empty stub when it is first called.

TL;DR: I wouldn't quite say it is *impossible* to make a "deep deinit" work in Zig, but the implementation is wrought with pitfalls and difficult tradeoffs. So any realistic proposal would require a great amount of detail and attention to corner cases. And it probably still won't work in the end.

**Comment by Tetralux** [comment]
> First off, ArrayList(T) is a generic type, so to perform a deep deinit, it needs to be able to call .deinit() on T regardless of what T is, including builtin types, pointers, etc. This circumstance pretty much requires that all types -- builtin or custom -- be given a default implementation of deinit() by the compiler.

To be fair, I'm not quite so sure about this point.

You could just have `deinit`'s implementation be:
```zig
pub fn deinit(self: *Self) void {
    if (@hasDecl(T, "deinit") and std.meta.isFn(T.deinit)) {
        for (self.items) |*e| e.deinit();
    }
    // free allocatedSlice(), etc.
}
```
You could then also provide a `deinitLeavingElements` which does what `deinit` does currently.
Or the other way round, so that `deinit` does as it currently does, but then add `deinitAndElements`, etc, etc.

**Comment by zzyxyzz** [comment]
I've put up a proposal (#8005) for how to solve this issue in a minimally intrusive manner. Thanks, @Tetralux.
